### DISCUSSION TITLE
Weakly coupling of two elements

### DISCUSSION THREAD

**Discussion by mnabideltares** [KratosMultiphysics/Kratos#10729] [Q&A]
Hi,

We would like to add a thermal element and couple it with the other processes. In our case, we couple the thermal element with water pressure  and deformation. 

We know it is possible to make an element which include degrees of freedom for all those variables (including temperature) and solve it in a fully coupled system. However, different combinations leads to defining of many type of elements (eg. Temperature-deformation, temperature-pressure, temperature-deformation-pressure, temperature alone, etc.) which we try to avoid. So, we are interested to use our thermal element, by means of weakly coupling with other elements. The temperature needs to communicate with the other element in iteration level.

I will be very thankful if somebody can tell us, whether it is possible, and how to make this coupling?

**Comment by loumalouomega** [comment]
You can do it using the CoSimulationApplication, which works as a black-box, and it is relatively easy to use when coupling Kratos-Kratos.

**Reply by mnabideltares** [reply]
Thank you @loumalouomega 
Is there any documentation available, explaining how to do it in CoSimulationApplication?

**Reply by loumalouomega** [reply]
The [README.md](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/CoSimulationApplication) includes plenty of documentation, and you can see the examples from: https://github.com/KratosMultiphysics/Examples/tree/master/co_simulation

**Reply by mnabideltares** [reply]
Thank you

**Comment by rubenzorrilla** [comment]
Alternatively, you can create a Python solver coupling two "sub"-solvers (see what is done in `coupled_fluid_thermal_solver.py`).

**Reply by mnabideltares** [reply]
@rubenzorrilla 
It is maybe the thing I was looking for.
I will try to implement a such solver for our purpose.
Thank you.

**Reply by rubenzorrilla** [reply]
OK. Just let us know if any problem raises.

**Reply by mnabideltares** [reply]
What I see from this script, it runs 2 processes from 2 different applications. However, I have only one application (GeoMechanics) and I want to use 2 different elements and couple them in iteration level. It is to avoid definition of many combinations in elements (avoids creating new elements).
I wonder whether this script is also suitable for a such purpose?
Thanks

**Reply by rubenzorrilla** [reply]
Sorry but I don't get your point at all... with "coupling at iteration level" you mean doing an external non-linear loop to couple both problems or to couple at the strategy (i.e. Newton-Raphson) level?

**Reply by mnabideltares** [reply]
They are physically coupled. When we decouple them by means of iteration, we need to couple them in iteration level. So, I mean couple at the strategy level, for each inner iteration.

**Reply by rubenzorrilla** [reply]
This sounds similar to a fractional step scheme. If so, you need to create an element implementing both physics (this can have as template arguments the two "parent" elements so you may don't need to re-implement everything) together with a custom strategy.

As you might guess, this is a rather complex implementation. Having said so and in lack of any details, I'd consider linearizing the temperature as a first try and see how it performs (i.e. to weakly couple the temperature to the mechanical problem). Nice thing of this is that it can be easily done in Python by following any of the suggestions above.

**Reply by mnabideltares** [reply]
I am trying to write a script like "coupled_fluid_thermal_solver.py", but calling the GeoMechanicsApplication twice, once for pressure and the other time is for the thermal part. Do you think it is possible to couple our two elements (both in GeoMechanicsApplication) by this way?

Is there any example (input files, json, mpda ... ) for coupling fluid-thermal?
Thank you for help.

**Reply by rubenzorrilla** [reply]
Yes, you can do it. However this turns into a weakly coupled solution strategy, which is not the same you were suggesting before ("to couple them in the iteration level").

There are no examples of the buoyancy solver in the Examples repo. However, you can easily create one from the already built ones in the GiD interface. You just need to select Thermic->Buoyancy->2D and then click in the examples button (the one with the pencil and the folder in the left vertical bar together to the tree). The example in there uses the `coupled_fluid_thermal_solver.py`.
